LocalLang/TypingGuarantees.lean

```diff
@@ -2,17 +2,198 @@ import LocalLang.AST
 import LocalLang.Semantics
 import LocalLang.Types
 import LocalLang.Typing
+import LocalLang.TypingLemmas
 
 def EnvRespectsCtx (Γ : TypeContext) (V : Env) : Prop := sorry
 
 -- TODO: prove
--- TODO: V and Γ should be related
 theorem preservation (Γ : TypeContext) (e e' : Expr) (ty : LLType)
     (h_env_respects : EnvRespectsCtx Γ V)
   : Expr.TypeJdg Γ e ty → HeadSmallStep V e e' → Expr.TypeJdg Γ e' ty := sorry
 
 -- Progress: if TypeJdg Γ e ty, then e can take a step.
 -- TODO: formalize
 
+def Expr.simple_rec.{l} {motive : Expr → Sort l}
+     (value_case : ∀ v, motive (.value v))
+     (const_case : ∀ n, motive (.const n))
+     (var_case : ∀ x, motive (.var x))
+     (bin_op_case : ∀ op lhs rhs, motive lhs → motive rhs → motive (.binOp op lhs rhs))
+     (let_in_case : ∀ name e₁ e₂, motive e₁ → motive e₂ → motive (.letIn name e₁ e₂))
+     (fun_case : ∀ f es, motive f → (∀ e', e' ∈ es → motive e') → motive (.funCall f es))
+    : (e : Expr) → motive e := go
+where go
+  | .value v => value_case v
+  | .const n => const_case n
+  | .var x => var_case x
+  | .binOp op lhs rhs => bin_op_case op lhs rhs (go lhs) (go rhs)
+  | .letIn name e₁ e₂ => let_in_case name e₁ e₂ (go e₁) (go e₂)
+  | .funCall f es => fun_case f es (go f) (fun e' _ => go e')
+
+inductive IsValue : Expr → Prop where
+  | value : IsValue (.value v)
+
+inductive IsRedex : Expr → Prop where
+  | const : IsRedex (.const n)
+  | var : IsRedex (.var x)
+  | binOp {op} : IsRedex (.binOp op (.value v₁) (.value v₂))
+  | letInConst : IsRedex (.letIn x (.value v₁) (.value v₂))
+  | funCall {es} : IsRedex (.funCall (.value v) es)
+
+inductive HasDecomposition : Expr → Prop where
+  | decomposition (e' : Expr) (ctx : Ctx) : e = ctx.fill e' → IsRedex e' → HasDecomposition e
```

Comment:
By the way, if you find yourself pattern matching on this in inconvenient ways, you may want to make it a structure instead.